core/domain/learner_goals_services.py

```diff
@@ -0,0 +1,140 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Services for the learner goals feature of the learner dashboard."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from core.domain import user_domain
+from core.platform import models
+import feconf
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+
+def get_learner_goals_from_model(learner_goals_model):
+    """Returns the learner goals domain object given the learner goals
+    model loaded from the datastore.
+
+    Args:
+        learner_goals_model: LearnerGoalsModel. The
+            learner goals model from the datastore.
+
+    Returns:
+        LearnerGoals. The learner goals domain object corresponding to the
+        given model.
+    """
+    return user_domain.LearnerGoals(
+        learner_goals_model.id,
+        learner_goals_model.topic_ids_to_learn,
+        learner_goals_model.topic_ids_to_master)
+
+
+def save_learner_goals(learner_goals):
+    """Save a learner goals domain object as an LearnerGoalsModel entity
+    in the datastore.
+
+    Args:
+        learner_goals: LearnerGoals. The learner goals domain object to
+            be saved in the datastore.
+    """
+    learner_goals_dict = learner_goals.to_dict()
+
+    learner_goals_model = (user_models.LearnerGoalsModel.get(
+        learner_goals.id, strict=False))
+    if learner_goals_model is not None:
+        learner_goals_model.populate(**learner_goals_dict)
+        learner_goals_model.update_timestamps()
+        learner_goals_model.put()
+    else:
+        learner_goals_dict['id'] = learner_goals.id
+        user_models.LearnerGoalsModel(**learner_goals_dict).put()
+
+
+def mark_topic_to_learn(user_id, topic_id):
+    """Adds the topic id to the learner goals of the user. If the count exceeds
+    feconf.MAX_CURRENT_GOALS_COUNT, the topic is not added.
+
+    Args:
+        user_id: str. The id of the user.
+        topic_id: str. The id of the topic to be added to the
+            learner goals.
+
+    Returns:
+        bool. The boolean indicates whether the learner goals limit
+        of the user has been exceeded.
+    """
+    learner_goals_model = user_models.LearnerGoalsModel.get(
+        user_id, strict=False)
+    if not learner_goals_model:
+        learner_goals_model = (
+            user_models.LearnerGoalsModel(id=user_id))
+
+    learner_goals = get_learner_goals_from_model(
+        learner_goals_model)
+
+    goals_limit_exceeded = False
+    topic_ids_count = len(learner_goals.topic_ids_to_learn)
+    if topic_id not in learner_goals.topic_ids_to_learn:
+        if topic_ids_count < feconf.MAX_CURRENT_GOALS_COUNT:
+            learner_goals.add_topic_id_to_learn(topic_id)
+        else:
+            goals_limit_exceeded = True
+    save_learner_goals(learner_goals)
```

Comment:
Are we going to make a save call even if the given topic already exists in `learner_goals.topic_ids_to_learn`? I would suggest raising exception!

Reply:
Done.